packages/medusa/src/utils/get-query-config.ts

```diff
@@ -19,93 +20,116 @@ export function pickByConfig<TModel extends BaseEntity>(
-  defaultRelations: string[],
-  fields?: (keyof TModel)[],
-  expand?: string[]
-): FindConfig<TModel> {
-  let includeFields: (keyof TModel)[] = []
-  if (isDefined(fields)) {
-    includeFields = Array.from(new Set([...fields, "id"])).map((field) => {
-      return typeof field === "string" ? field.trim() : field
-    }) as (keyof TModel)[]
-  }
-
-  let expandFields: string[] = []
-  if (isDefined(expand)) {
-    expandFields = expand.map((expandRelation) => expandRelation.trim())
-  }
-
-  return {
-    select: includeFields.length ? includeFields : defaultFields,
-    relations: isDefined(expand) ? expandFields : defaultRelations,
-  }
-}
+export function prepareListQuery<
+  T extends RequestQueryFields,
+  TEntity extends BaseEntity
+>(validated: T, queryConfig: QueryConfig<TEntity> = {}) {
+  const { order, fields, limit = 50, expand, offset = 0 } = validated
+  let {
+    allowed = [],
+    defaults = [],
+    defaultFields = [],
+    defaultLimit,
+    allowedFields = [],
+    allowedRelations = [],
+    defaultRelations = [],
+  } = queryConfig
+
+  allowedFields = allowed.length ? allowed : allowedFields
+  defaultFields = defaults.length ? defaults : defaultFields
+
+  // Fields such as *variants.options meaning that we want to select all fields
+  // from the variants.options relation without specifying them.
+  // In that case it needs to be removed from the fields and be part of the relations only
+  // TODO: It means that the remote query won't have it as part of the fields
+  // unless the properties are specified e.g variants.options.id, variants.options.name
+  // This is a limitation of the current implementation as we do not support select * from a
+  // relation. Once we add the support for it then we can update this logic
+  const starFields = new Set()
+
+  let allFields = new Set(defaultFields) as Set<string>
 
-export function getListConfig<TModel extends BaseEntity>(
-  defaultFields: (keyof TModel)[],
-  defaultRelations: string[],
-  fields?: (keyof TModel)[],
-  expand?: string[],
-  limit = 50,
-  offset = 0,
-  order: { [k: string | symbol]: "DESC" | "ASC" } = {}
-): FindConfig<TModel> {
-  let includeFields: (keyof TModel)[] = []
   if (isDefined(fields)) {
-    const fieldSet = new Set(fields)
-    // Ensure created_at is included, since we are sorting on this
-    fieldSet.add("created_at")
-    fieldSet.add("id")
-    includeFields = Array.from(fieldSet) as (keyof TModel)[]
-  }
+    const customFields = fields.split(",").filter(Boolean)
+    const shouldReplaceDefaultFields =
+      !customFields.length ||
+      customFields.some(
+        (field) => !(field.startsWith("-") || field.startsWith("+"))
+      )
+    if (shouldReplaceDefaultFields) {
+      allFields = new Set(customFields.map((f) => f.replace(/^[+-]/, "")))
+    } else {
+      customFields.forEach((field) => {
+        if (field.startsWith("+")) {
+          allFields.add(field.replace(/^\+/, ""))
+        } else if (field.startsWith("-")) {
```

Comment:
I guess replace will only replace the first occurrence, but we need to be careful not to mess up a variable name called `some-variable` for example. Just something that can easily happen 😄

Reply:
yeah exactly, but I can easily see someone coming in later on and changing the code thereby breaking the logic :D Just having a test for it would be nice.

Reply:
@carlos-r-l-rodrigues @sradevski you ll have to re review, I ve done some changes in the mean time and the test includes this check